rules/firestore.json

```diff
@@ -35,50 +37,204 @@ service cloud.firestore {
+            // Moreover, the username cannot be changed once set
+            // And the verified field can only be changed to true if the user has a username and has its email verified
+            allow update: if isOwner(userId) && doesNotModifyUsernameIfSet() && onlyModifiesVerifiedFieldIfUserIsVerified();
+            // Anybody can read the fields of a user.
+            // This is because before user creation, we need to check if the username is free.
             allow read: if true;
         }
-        match /parties/{id} {
-            allow write, read: if userVerified();
-        }
-        match /joiners/{id} {
-            allow write, read: if userVerified();
+        function isOwner(userId) {
+            return request != null &&
+                request.auth != null &&
+                request.auth.uid == userId;
         }
-        function userVerified() {
-            return isAuthentified() &&
+        function doesNotModifyUsernameIfSet() {
+            let usernameAlreadySet = hasField(resource.data, 'username');
+            let usernameBeingSet = hasField(request.resource.data, 'username');
+            // It sometimes happens that firebase sends the username as part as updates, even though it is not actually updated. Hence, we accept usernames as part of updates, only if it doesn't change.
+            let doesNotChange = usernameAlreadySet && usernameBeingSet && resource.data['username'] == request.resource.data['username'];
+            return usernameAlreadySet == false || usernameBeingSet == false || doesNotChange;
+        }
+        function onlyModifiesVerifiedFieldIfUserIsVerified() {
+            let verifiedKeptSameValue = (resource.data.verified == request.resource.data.verified);
+            let verifiedSetToTrueAfterVerification =  (resource.data.verified == false && userIsVerified() && resource.data.username != null);
+            return verifiedKeptSameValue || verifiedSetToTrueAfterVerification
+        }
+        match /parts/{id} {
+            // Any verified user can create a part with the proper fields set, and has to be the creator
+            allow create: if userIsVerified() &&
+                             // turn can be -1 in case of a new part, 0 in case of a rematch
+                             (request.resource.data.turn == -1 || request.resource.data.turn == 0) &&
+                             hasField(request.resource.data, 'typeGame') &&
+                             request.resource.data.listMoves == [] &&
+                             // creator must be one of the players (could be playerOne in case of rematch)
+                             (request.auth.uid == request.resource.data.playerZero.id ||
+                              hasField(request.resource.data, 'playerOne') && request.auth.uid == request.resource.data.playerOne.id);
+            // The owner of a part can delete it, if it has not started
+            allow delete: if userIsVerified() &&
```

Comment:
à chaque requête ce script est lu du début à la fin right ?
On peut pas constantiser userIsVerified au début, ça diminuerais po mal de métriques non ? On garde ça pour plutarque ?

Reply:
euh constantiser comment ?
La sémantique est: si tu fais un delete par exemple, il regarde quelle collection matche, et regarde les règles allow qui correspondent à la requête. S'il trouve *un seul* endroit où le `allow` est évalué à `true`, le delete est fait. Donc c'est un gros OR, et faut dupliquer pas mal

Reply:
mmh, il checke les delete même quand le "match /" ne matche pas ?

Ce qui coûte concrêtement c'est quoi, pas le nombre de fonction qu'on a mais plutôt genre les lectures de docs ? c'pas accéder à request et resource, si :p ?

Reply:
Comment ça il check les delete ? Si aucun match ne retourne true, la requête est refusée

Ce qui coûte c'est effectivement les requêtes (`exists`, `get`), mais pas les accès aux variables accessibles (`resource`)

Reply:
je veux dire, est ce que les allow <operation> if <condition> sont bel et bien ignorés si le "match /" dans lequel ils sont imbriqués est faux ? Il me semble super évident que non mais c'est pour être sûr :p 

Quand a constantiser je voulais dire c'est que pour être sur que no exists/get sont appellés qu'une fois on peut faire un système pour s'assurer qu'on demande qu'une fois à firebase puis qu'on le réutilise les fois suivantes (par exemple un optional qu'est vide tant qu'on a pas checké)

Reply:
ah, oui effectivement ! 

Pour la constantisation, à nouveau: le langage pue la mort, mais je crois qu'il optimise, ça m'étonnerait pas qu'il fassent les requêtes qu'une seule fois. Dans tous les cas, ça sera pour le ticket optimisation je dirais :p

Reply:
en plus d'inliner, avoir des stat de qu'est ce qui est le plus souvent faux pourrait être fun, comme ça on le met en premier dans les AND et ce qui est le plus souvent vrai dans les OR !

Reply:
oui pour la v1000 :p